packages/vitest/src/node/watch-filter.ts

```diff
@@ -0,0 +1,163 @@
+
+  public async filter(initial: string, filterFunc: FilterFunc): Promise<string | undefined> {
+    stdout().write(`${ESC}?25h`)
+    stdout().write(`${c.cyan('?')} ${c.bold(this.message)} › `)
+    return new Promise((resolve) => {
+      this.onKeyPress = this.filterHandler(filterFunc, (result) => {
+        this.close()
+        resolve(result)
+      })
+      process.stdin.on('keypress', this.onKeyPress)
+    })
+  }
+
+  private close() {
+    this.filterRL.close()
+    if (this.onKeyPress)
+      process.stdin.removeListener('keypress', this.onKeyPress)
+
+    if (process.stdin.isTTY)
+      process.stdin.setRawMode(false)
+  }
+
+  private restoreCursor() {
+    const cursortPos = this.keywordOffset() + (this.currentKeyword?.length || 0)
+    stdout().write(`${ESC}${cursortPos}G`)
+  }
+
+  private cancel() {
+    stdout().write(`${ESC}J`) // erase down
+  }
+
+  private filterHandler(filterFunc: FilterFunc, onSubmit: (result?: string) => void) {
+    return async (str: string, key: any) => {
+      switch (true) {
+        case key.sequence === '\x7F':
+          if (this.currentKeyword && this.currentKeyword?.length > 1)
+            this.currentKeyword = this.currentKeyword?.slice(0, -1)
+
+          else
+            this.currentKeyword = undefined
+
+          break
+        case key?.ctrl && key?.name === 'c':
+        case key?.name === 'escape':
+          this.cancel()
+          onSubmit(undefined)
+          break
+        case key?.name === 'enter':
+        case key?.name === 'return':
+          onSubmit(this.results[this.selectionIndex] || this.currentKeyword || '')
+          this.currentKeyword = undefined
+          break
+        case key?.name === 'up':
+          if (this.selectionIndex && this.selectionIndex > 0)
+            this.selectionIndex--
+          else
+            this.selectionIndex = -1
+
+          break
+        case key?.name === 'down':
+          if (this.selectionIndex < this.results.length - 1)
+            this.selectionIndex++
+          else if (this.selectionIndex >= this.results.length - 1)
+            this.selectionIndex = this.results.length - 1
+
+          break
+        case !key?.ctrl && !key?.meta:
+          if (this.currentKeyword === undefined)
+            this.currentKeyword = str
+
+          else
+            this.currentKeyword += str
+          break
+      }
+
+      if (this.currentKeyword)
+        this.results = await filterFunc(this.currentKeyword)
+
+      this.render()
+    }
+  }
+
+  private render() {
+    if (this.currentKeyword) {
+      this.printKeyword()
+
+      if (this.results.length === 0) {
+        this.eraseAndPrint(`\nPattern matches no results`)
+      }
+      else {
+        const resulCountLine = `Pattern matches ${this.results.length} results`
+
+        let resultBody = ''
+
+        if (this.results.length > MAX_RESULT_COUNT) {
+          const offset = this.selectionIndex > SELECTION_MAX_INDEX ? this.selectionIndex - SELECTION_MAX_INDEX : 0
```

Comment:
implement scroll logic for long results.

## demo

https://github.com/vitest-dev/vitest/assets/19663115/1d4aefa6-3262-4e04-912c-9f9b52710706